### PR TITLE
rados.py: wrap omap operation for librados

### PR DESCRIPTION
**Pull request by chenji-kael**
you can operate omap in python now

Signed-off-by: chenji <insomnia@139.com>

### CODE REVIEW

**Comment by chenji-kael** [comment]
@tchaikov Thanks for your suggestion, I have fix it

File: src/pybind/rados.py [hunk 1]
```diff
@@ -1947,6 +2161,13 @@ def retfunc(self, *args, **kwargs):
         return retval
     return retfunc
 
+class OmapItem(object):
+    def __init__(self, key, value, lens):
+        self.key = key
+        self.value = value
+        self.lens = lens
+    def __str__(self):
+        return "Omap.Iteam(key=%s, value=%s, lens=%d)" % (self.key, self.value, self.lens)
```
Comment (tchaikov) [inline src/pybind/rados.py]: typo ? s/Iteam/Item/

File: src/pybind/rados.py [hunk 2]
```diff
@@ -1821,16 +1847,204 @@ def lookup_snap(self, snap_name):
 
     def get_last_version(self):
         """
-        Return the version of the last object read or written to.
+            Return the version of the last object read or written to.
```
Comment (tchaikov) [inline src/pybind/rados.py]: i'd suggest put this change into another commit. but could you enlighten me why you want to indent this doc string? seems we don't indent them in other places.

File: src/pybind/rados.py [hunk 3]
```diff
@@ -1821,16 +1847,204 @@ def lookup_snap(self, snap_name):
...
+           :para keys :a tuple of keys
+           :type tuple
+           :para values :a tuple of values
+           :type tuple
+           :para value_lens :lens of values
+           :type tuple
+           :para count: total counts of key-value pairs
+        """
+        if len(keys) != len(values):
+            raise Error("Rados(): keys and values must hava the same numbers")
```
Comment (tchaikov) [inline src/pybind/rados.py]: typo s/hava/have/

and i'd suggest put in this way

> keys and values must have the same number of items.

File: src/pybind/rados.py [hunk 4]
```diff
@@ -1821,16 +1847,204 @@ def lookup_snap(self, snap_name):
...
+        """
+           excute the real read operation
+           :para read_op :read operation object
+           :type int
+           :para oid : object name
+           :type str
+           :para flags:flags to apply to the entire operation
+           :type int
+        """
+        if not isinstance(oid, str):
```
Comment (tchaikov) [inline src/pybind/rados.py]: could dispense this check now.

File: src/pybind/rados.py [hunk 5]
```diff
@@ -1821,16 +1847,204 @@ def lookup_snap(self, snap_name):
...
+           get the omap values
+           :para read_op :read operation object
+           :type int
+           :para start_after : list keys starting after start_after
+           :type str
+           :para filter_prefix: list only keys beginning with filter_prefix
+           :type int
+           :para max_return:list no more than max_return key/value pairs
+           :type int
+           return:object(OmapIterator),return value from this action
```
Comment (tchaikov) [inline src/pybind/rados.py]: should be `:returns:`

and 
> object(OmapIterator),return value from this action

does not help much. probably:

> an iterator over the the requested omap values

File: src/pybind/rados.py [hunk 6]
```diff
@@ -1821,16 +1847,204 @@ def lookup_snap(self, snap_name):
...
+        key_array[:] = keys
+        run_in_thread(self.librados.rados_write_op_omap_rm_keys,
+                      (c_void_p(write_op), byref(key_array), c_int(key_num)))
+    
+    @requires(('write_op', int))
+    def clear_omap(self, write_op):
+        """
+           Remove all key/value pairs from an object
+           :para write_op :write operation object
+           :type int
```
Comment (tchaikov) [inline src/pybind/rados.py]: should be `:type write_op:`

File: src/pybind/rados.py [hunk 7]
```diff
@@ -1821,16 +1847,204 @@ def lookup_snap(self, snap_name):
...
+        key_array = key_array_type()
+        key_array[:] = keys
+        run_in_thread(self.librados.rados_write_op_omap_rm_keys,
+                      (c_void_p(write_op), byref(key_array), c_int(key_num)))
+    
+    @requires(('write_op', int))
+    def clear_omap(self, write_op):
+        """
+           Remove all key/value pairs from an object
+           :para write_op :write operation object
```
Comment (tchaikov) [inline src/pybind/rados.py]: i am not sure if the ending space would help. please see http://sphinx-doc.org/domains.html#info-field-lists for more details on how we document the python code.

**Comment by tchaikov** [comment]
just put some cosmetic comments, will look closer later on.

File: src/pybind/rados.py (hunk 2 above)
Comment (chenji-kael) [inline src/pybind/rados.py]: er,i think this a type error,i will change it back

File: src/pybind/rados.py (hunk 3 above)
Comment (chenji-kael) [inline src/pybind/rados.py]: i want make the input more simple ,so i have count these two para in function,but forget to delete the comment

File: src/pybind/rados.py (hunk 7 above)
Comment (chenji-kael) [inline src/pybind/rados.py]: Thanks for your help,I am new to pathon ,and will follow your advise :)

**Comment by chenji-kael** [comment]
@tchaikov ping 
hi,tchaikov , I have a new commit , could you help me to check? Thanks a lot!

File: src/pybind/rados.py [hunk 8]
```diff
@@ -831,6 +831,32 @@ def blacklist_add(self, client_address, expire_seconds=0):
...
+        :returns: next rados.OmapItem
+        """
+        self.key = c_char_p()
+        self.value = c_char_p()
+        self.lens = c_int()
+        run_in_thread(self.ioctx.librados.rados_omap_get_next,
+                      (self.ctx, pointer(self.key), pointer(self.value),
+                       pointer(self.lens)))
+        if isinstance(self.value.value, str):
+            self.value.value = self.value.value[0:self.lens.value]
```
Comment (tchaikov) [inline src/pybind/rados.py]: probably should raise `StopIteration` in such a case `if self.value.value is None`?

File: src/pybind/rados.py [hunk 9]
```diff
@@ -1947,6 +2156,13 @@ def retfunc(self, *args, **kwargs):
         return retval
     return retfunc
 
+class OmapItem(object):
+    def __init__(self, key, value, lens):
+        self.key = key
+        self.value = value
+        self.lens = lens
```
Comment (tchaikov) [inline src/pybind/rados.py]: could you add an empty line between functions? see https://www.python.org/dev/peps/pep-0008/#blank-lines

**Comment by tchaikov** [comment]
@chenji-kael you might want to squash all these three commit into a single one? thanks.

File: src/pybind/rados.py (hunk 8 above)
Comment (chenji-kael) [inline src/pybind/rados.py]: yeah,that`s right

**Comment by chenji-kael** [comment]
@tchaikov ping hi, only one commit remains now, thanks for your guidance

File: src/test/pybind/test_rados.py [hunk 10]
```diff
@@ -337,6 +337,72 @@ def test_remove_snap(self):
...
+        self.ioctx.release_read_op(read_op)
+
+    def test_clear_omap(self):
+        keys = ("1", "2", "3")
+        values = ("aaa", "bbb", "ccc")
+        write_op = self.ioctx.create_write_op()
+        self.ioctx.set_omap(write_op, keys, values)
+        self.ioctx.operate_write_op(write_op, "hw")
+        self.ioctx.clear_omap(write_op)
+        self.ioctx.operate_write_op(write_op, "hw")
```
Comment (jdurgin) [inline src/test/pybind/test_rados.py]: it's not generally safe to use the same write_op for more than one operate_write_op() call. The same goes for read ops. It'd be good to have a test for multiple ops in the same write_op here though.

**Comment by jdurgin** [comment]
Looks good in general, thanks! Just a small note in one of the tests

**Comment by chenji-kael** [comment]
@tchaikov hi, I update it again , thank you!

**Comment by tchaikov** [comment]
@chenji-kael we are close. just run the test in my testbed. some of them failed. once you get them addressed, i think this change will get merged. thanks.

**Comment by tchaikov** [comment]
@chenji-kael ping?

**Comment by chenji-kael** [comment]
HI, I am coding right now ,maybe tomorrow, I will give a new commit after I pass the unit test

**Comment by tchaikov** [comment]
@chenji-kael thanks. not rushing you. just want to make sure i am not blocking you. =)

**Comment by chenji-kael** [comment]
@tchaikov hi,I got a new commit, and create two new classes: maybe their names are not proper ↖(▔▽▔)↗
class ReadOpCtx(object):
class WriteOpCtx(object):

**Comment by tchaikov** [comment]
@chenji-kael neat! the tests passed. and once the leak is fixed, we are good to merge!

> maybe their names are not proper 

`{Read,Write}OpCtx` are good. probably we can turn them into fully functional `{Read,Write}Op` in future by adding a proper constructor and a `release()` method. (think about the "file" object in python).

**Comment by chenji-kael** [comment]
@tchaikov leak has been fix!

**Comment by tchaikov** [comment]
@chenji-kael are you sure you want to have "chenji" as your real name, per https://github.com/ceph/ceph/blob/master/SubmittingPatches#L62 we'd like have the real name of the contributor in the `Signed-off-by` line .

thought it would be "Ji Chen". if your first name is "Ji"

**Comment by tchaikov** [comment]
`Reviewed-by: Kefu Chai <kchai@redhat.com>`  once get the nod from @chenji-kael

**Comment by chenji-kael** [comment]
@tchaikov yeah,my family name is Chen,so “Ji Chen” is the standard English expression，"Ji Chen" is better.

**Comment by tchaikov** [comment]
@chenji-kael could you update your commit message then? it is ended with:
```
Signed-off-by: chenji <insomnia@139.com>
```

**Comment by chenji-kael** [comment]
@tchaikov done
